### ISSUE TITLE
[New]: Add a S-parameter calculation helper

### ISSUE THREAD

**Issue by QuentinWach** [QuentinWach/beamz#71]
Go from GDSFactory design to S-parameter.

<img width="771" height="117" alt="Image" src="https://github.com/user-attachments/assets/bf28c3bf-95a9-4a3e-9787-4f677f93325e" />

ChatGPT:

### Port & mode labeling

GDSFactory’s plugin ecosystem uses port+mode labels like `o1@0,o2@0` (“port o1 mode 0 to port o2 mode 0”). ([[gdsfactory.github.io](https://gdsfactory.github.io/gplugins/plugins_fdtd.html?utm_source=chatgpt.com)][1])
That’s a good convention to adopt in BEAMZ because:

* it scales to multimode / polarization without inventing new concepts
* it aligns with existing tooling (plotters, converters, stored npz conventions)

### SAX expects a callable “model”

SAX’s philosophy is: **you give it a function (often JAX-friendly) that returns S-parameters**; it tries to stay thin around JAX and not impose heavy data structures. ([[flaport.github.io](https://flaport.github.io/sax/?utm_source=chatgpt.com)][2])
So your end product should be:

```python
def model(wl_um: jnp.ndarray, **params) -> SDict:
    ...
    return sdict
```

(where `SDict` is basically a dict keyed by `(port_in, port_out)` or string keys depending on your adapter.)

---

## 2) Layout → simulation geometry (BEAMZ side)

You need one canonical adapter that turns a `gf.Component` into a BEAMZ simulation domain:

**Inputs**

* `component` (GDSFactory)
* `layer_stack` / thicknesses / materials
* `background/clad` materials
* `ports` (which ones are “circuit ports”)
* `port extensions` (straight sections for mode settling)
* `mesh / dx`
* `wavelength grid`

**Outputs**

* A BEAMZ “SimSpec” that contains:

  * geometry/epsilon function (your SDF pipeline is perfect here)
  * port planes (locations + normals + cross-section)
  * requested modes per port (TE0/TM0 or mode indices)

If you mirror gplugins conceptually, you’ll feel familiar to users coming from `gdsfactory + gplugins` (tidy3d/meep/lumerical notebooks do exactly “simulate each component’s S-params then solve the circuit in SAX”). ([[gdsfactory.github.io](https://gdsfactory.github.io/ubc/notebooks/12_sim_plugins_tidy3d.html?utm_source=chatgpt.com)][3])

---

## 3) Simulation → S-matrix (per wavelength)

For each wavelength (or a small batch of wavelengths if you do broadband time-domain + FFT):

1. **Solve port modes** at each port cross-section (at that wavelength)
2. For each *input port/mode*:

   * excite that mode (normalized to 1 W or 1 a.u. consistently)
   * run to steady state / collect fields at monitors
   * **project outgoing fields onto port modes** at every port to get complex amplitudes
3. Fill `S[out_port@j, in_port@i]` for all ports/modes

Key implementation detail: **be strict about normalization** (power normalization of modes + consistent reference planes). If you get this right once, everything downstream becomes boring.

---

## 4) Store results in a circuit-friendly artifact

You want two storage forms:

### A) “Raw” sampled data

* `wavelengths`
* `S[w, out, in]` complex
* metadata: port names, mode indices, reference plane offsets, units

### B) A “model” that SAX can call

SAX workflows typically want a callable that returns an SDict across wavelengths. (GDSFactory even has helpers to plot a “model function” like that.) ([[gdsfactory.github.io](https://gdsfactory.github.io/gplugins/_autosummary/gplugins.sax.plot_model.html?utm_source=chatgpt.com)][4])

You have two options:

**Option 1 (simplest): interpolation model**

* store sampled S over wavelength
* return values via linear/spline interpolation

**Option 2 (best for speed + stability): rational fit**

* vector-fit each Sij(w)
* evaluate fast at arbitrary wavelengths
* nicer gradients if you do it carefully (or treat fit coeffs as constants)

Start with interpolation; add rational fitting later.

---

### User-facing “one-liner”

```python
import beamz as bz

sp = bz.sparameters.from_gdsfactory(
    component=c,
    layer_stack=LS,
    wavelengths_um=jnp.linspace(1.50, 1.60, 201),
    ports="optical",          # default: all optical ports
    modes=[0],                # default: fundamental
    boundary="pml",
    mesh=bz.Mesh(dx=20e-9),
)

model = sp.to_sax_model(interp="linear")   # returns callable(wl)->SDict
```

### Cached + deterministic by default

Under the hood, make the simulation cache key depend on:

* component hash (gds + ports + layerstack)
* sim settings (dx, pml, solver settings)
* wavelength grid

This matches the expectation people already have in gdsfactory simulation plugins: “if it was simulated once, reuse it.” ([[gdsfactory.github.io](https://gdsfactory.github.io/gplugins/notebooks/lumerical_1_fdtd_sparameters.html?utm_source=chatgpt.com)][5])

### A “component model registry” hook (optional but powerful)

If you want it to feel *PDK-like*:

```python
bz.models.register("mmi1x2", model)
# or attach to gf.Component.info["model"] / metadata pattern
```

---

## 6) How it plugs into SAX

At the end, users should hand SAX a dict mapping instance names to model functions (that’s the common pattern in SAX-adjacent tooling). ([[piel.readthedocs.io](https://piel.readthedocs.io/en/latest/examples/03_sax_basics.html?utm_source=chatgpt.com)][6])

Conceptually:

```python
models = {
  "mmi": model_mmi,     # callable(wl)->SDict
  "bend": model_bend,
}
# then sax solves the connected network from those building blocks
```

(Exact circuit assembly depends on whether they use pure `sax` or something like simphony/piel; but the **model function contract** is the stable piece.)

---

## 7) What I’d implement first (so it’s not a science project)

1. **Port/mode extraction + normalization** (solid, tested)
2. **S-parameter extraction for 2-port and 4-port components**
3. **NPZ-based cache format**
4. **`to_sax_model()` with interpolation**
5. Only then: rational fitting, multimode/pol, sensitivity/gradients

[1]: https://gdsfactory.github.io/gplugins/plugins_fdtd.html?utm_source=chatgpt.com "FDTD — GDSFactory Plugins"
[2]: https://flaport.github.io/sax/?utm_source=chatgpt.com "sax"
[3]: https://gdsfactory.github.io/ubc/notebooks/12_sim_plugins_tidy3d.html?utm_source=chatgpt.com "Circuit simulations (tidy3d + SAX) — ubc"
[4]: https://gdsfactory.github.io/gplugins/_autosummary/gplugins.sax.plot_model.html?utm_source=chatgpt.com "gplugins.sax.plot_model"
[5]: https://gdsfactory.github.io/gplugins/notebooks/lumerical_1_fdtd_sparameters.html?utm_source=chatgpt.com "FDTD Ansys Lumerical — GDSFactory Plugins"
[6]: https://piel.readthedocs.io/en/latest/examples/03_sax_basics.html?utm_source=chatgpt.com "SAX Integration Basics - piel - Read the Docs"
